### DISCUSSION TITLE
One-Off: new release

### DISCUSSION THREAD

**Discussion by zmontgo** [LeoBorai/mailjet-rs#58] [General]
Just wanted to propose you make a new release, v0.2.2, to include the new headers in the crates.io release until the API 3.1 update can be released.

**Comment by zmontgo** [comment]
Edit: I just read your last comment before you closed the pull request. Thank you much.

**Comment by LeoBorai** [comment]
Hi @zmontgo! Thanks for kickstarting this discussion!
As we have been discussing on #5, the Mailjet Send API 3.1 support is planned but is not yet started.

Our current architecture is intended for splitting the support for Send API V3 and Send API V3.1, and at first I thought it would be nice to have both implementations, but as of today I don't see any benefits on doing such approach.

I believe that we could actually drop support for Send API V3 and move into supporting versions from Send API V3.1 onwards.

What do you think about doing this for a next release?

**Comment by LeoBorai** [comment]
Hey @zmontgo hows going?

I have been checking other API Clients written in Rust looking for patterns/architecture, and I found very intuitive and simple to understand, the Cloudflare's API Client for Rust on [cloudflare/cloudflare-rs](https://github.com/cloudflare/cloudflare-rs).

Let me know your thoughts!

**Comment by zmontgo** [comment]
@EstebanBorai sorry for ghosting you, tumultuous time in my life. If you're still looking to work on this, let's do it! Otherwise, consider adding me as a maintainer and I'll start spearheading development. Up to you!

**Reply by LeoBorai** [reply]
Hi! No worries! Of course we can still work on this together!

**Comment by zmontgo** [comment]
Ideally we should offer a comprehensive, wrapper that encapsulates the entire API surface with feature flags/modular design. I can write clear docs along the way.

**Reply by LeoBorai** [reply]
Yeah! Feel free to open issues as you go, so we discuss each layer together!

**Comment by zmontgo** [comment]
I really love the way the Cloudflare API client looks. My proposal is, if we are already wanting to implement breaking changes, let's opt to support all versions of the API and rebuild it from the ground up to make some quality of life improvements. I still have some things to go through, but a few of the changes I want to make:

- Never panic
- Use a builder design pattern rather than mutable references
- Define the possible API responses and their structure using enums and structs, automatically parsing using `serde` [things like `#[serde(untagged)]` are especially helpful]
- Use the `reqwests` library
- Restructure and rename the file structure, as well as file/method names

Some of these were inspired by the `cloudflare-rs` package. I really love this...
```rust
#![forbid(unsafe_code)]

pub mod endpoints;
pub mod framework;
```

One other thing, just worth mentioning, is that I was seized with the idea of letting this compile into wasm, but I am not sure there would be enough consumers [due to the fact that you would have to send private API keys and let the client decide when to send emails] to make that officially supported. Maybe I'm not thinking of something, so if you can we should consider that when architecting this.

If I remember correctly, this was the direction you were wanting to go, and I'll be happy to co-author this with you. I think a method of communication more conducive to frequent back-and-forth may be useful if this is what we end up doing, so feel free to email me at `github.431bl@passmail.net` and I can give you my Discord username or whatever else you'd like.

**Reply by LeoBorai** [reply]
Sounds good to me! Very good insights and ideas @zmontgo!
I agree with your take on this, feel free to begin working on it!

**Comment by zmontgo** [comment]
Hey @LeoBorai! It's been a minute since I've followed up with you. I hope you're doing well.

I have been preoccupied with school, but now I am back and ready to get some work done. For one of my recent projects, I'm wanting a Mailjet crate that can also handle a typed builder pattern for MJML. I'm scaffolding some of the PoC with Claude (in spite of my... many reservations about AI code; it will serve to get something at least working for the time being, and then I can go in and maintain it manually).

I do not want to compete with you, but because the direction is so different, I am thinking about starting my own crate. I want to invite you to have equal authorship and maintenance authority over the project.

Feel free to look at the repo at https://github.com/zmontgo/manteau (I should be uploading the first version in a couple hours). If you're interested, please let me know, and I can give you more technical details.
